### PR TITLE
JS: Add support for JWT libraries

### PR DESCRIPTION
**Pull request by erik-krogh**

### CODE REVIEW

File: javascript/2020-11-09-jwt.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+lgtm,codescanning
+* Libraries implementing JWT have been modelled.
```
Comment (intrigus-lgtm) [inline javascript/2020-11-09-jwt.md:2]:
```suggestion
* Libraries implementing JWT have been modeled. 
```

File: javascript/ql/src/semmle/javascript/frameworks/JWT.qll [hunk 2]
```diff
@@ -0,0 +1,54 @@
+/**
+ * Provides classes for working frameworks implementation various aspects of JSON web tokens.
+ */
+
+import javascript
+
+/**
+ * Provides classes and predicates modelling the `jwt-decode` libary.
```
Comment (intrigus-lgtm) [inline javascript/ql/src/semmle/javascript/frameworks/JWT.qll:8]:
```suggestion
 * Provides classes and predicates modeling the `jwt-decode` libary.
```

File: javascript/ql/src/semmle/javascript/frameworks/JWT.qll [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+    override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
+      pred = this.getArgument(0) and
+      succ = this
+    }
+  }
+}
+
+/**
+ * Provides classes and predicates modelling the `jsonwebtoken` libary.
```
Comment (intrigus-lgtm) [inline javascript/ql/src/semmle/javascript/frameworks/JWT.qll:25]:
```suggestion
 * Provides classes and predicates modeling the `jsonwebtoken` libary.
```

File: javascript/2020-11-09-jwt.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+lgtm,codescanning
+* Libraries implementing JWT have been modeled.
```
Comment (asgerf) [inline javascript/2020-11-09-jwt.md:2]:
```suggestion
* The security queries now track taint through JWT decoding, and warns about hard-coded JWT signing keys.
```
I think we should try to be a little less opaque in the these change notes.

It should also be moved to the `javascript/change-notes/` directory.

File: javascript/ql/src/semmle/javascript/frameworks/JWT.qll [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    VerifyStep() { this = DataFlow::moduleMember("jsonwebtoken", "verify").getACall() }
+
+    override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
+      pred = this.getArgument(0) and
+      succ = this.getABoundCallbackParameter(2, 1)
+    }
+  }
+
+  /**
+   * The public/private key for a JWT as a `CredentialsExpr`.
```
[the remark this answers is not in the document]
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/frameworks/JWT.qll:41]: It can also use symmetric encryption.

[the remark this answers is not in the document]
Comment (asgerf) [reply javascript/ql/src/semmle/javascript/frameworks/JWT.qll:41]: I still think we should remove it from `verify`. If symmetric encryption is used we should still be able to pick it up the same key via the `sign` call.

**Review by asgerf — APPROVED** [review]

**Pull request merged**